### PR TITLE
Add @property information to Record class to make autocompletion easier

### PR DESCRIPTION
**Pull request by sunkan**
Add property to Record classes as a quality of life improvment

### CODE REVIEW

**Comment by pmjones** [comment]
Hi @sunkan -- I thought once about adding something like this as well; thanks!

However, I realized that if the table info changes, the table class will get overwritten, but the mapper class will not. This means that the mapper will be out of sync with the table, until the properties get hand-edited.

Have you considered finding a way to update *just* the property docblock portion in the mapper?

**Comment by pmjones** [comment]
Thinking about this some more, I had an idea:

1. Instead of adding docblocks on the Record itself, put them in a new "{Mapper}Fields" class. The class would be otherwise empty. The "{Mapper}Fields" class would extend the AbstractRecord.

2. Then the "{Mapper}Record" class can extend the "{Mapper}Fields" class.

3. Finally, the CLI can safely overwrite the "{Mapper}Fields class" since it's not the "real" Record class.

This would not break BC, either, as the new Fields class would get generated, but old Records would not already have extended it; they would continue to extend AbstractRecord until the developer manually changed them.

Does all that make sense? Interested to hear comments/criticism.

**Comment by sunkan** [comment]
Sound interesting. 

Have changed the code to format.

But instead of adding an empty class to the class hierarchy. 
We could use an Interface or a Trait.

File: src/Skeleton/Skeleton.php [hunk 1]
```diff
@@ -279,18 +279,22 @@ protected function setVars()
         $info = '[' . PHP_EOL . $info . '        ]';
 
         $cols = "[" . PHP_EOL;
+        $properties = PHP_EOL;
         $default = "[" . PHP_EOL;
         foreach ($list as $col => $val) {
             $val = ($val === null) ? 'null' : var_export($val, true);
             $cols .= "            '$col'," . PHP_EOL;
+            $properties .= " * @property string \$$col" . PHP_EOL;
```
Comment (froschdesign) [inline src/Skeleton/Skeleton.php]: I think the `null` value is missing.

**Comment by sunkan** [comment]
@froschdesign Updated code to reflect that

**Comment by pmjones** [comment]
NICE.

Last three things before I merge.

1. The PR is against the 1.x branch; do you really intend it for 1.x, or should it be for 2.x? (Or both?)

2. The @property type should probably be "mixed" instead of "string." It might be nice to append a description on the property that consists of the column definition, a la `@property mixed $foo VARCHAR(255) NOT NULL` -- I think some IDEs show the description.

3. Did you want to change this to a trait or interface, or you are you OK with the empty class after all? (I'm still OK with the class.)

This looks great. Thanks for your work here.

**Comment by sunkan** [comment]
1. There is no 2.x branch on Atlas.Cli

2. Doesn't PDO return all columns as string?

3. Doesn't really matter

**Comment by pmjones** [comment]
1. Oh of course.

2. It does, but they can get set to anything at all after that.

3. Good enough.

Merging!

**Comment by froschdesign** [comment]
@sunkan @pmjones 
> 2. Doesn't PDO return all columns as string?
> …
> 2. It does, but they can get set to anything at all after that.

You can change the behaviour: `PDO::ATTR_STRINGIFY_FETCHES`

http://php.net/manual/en/pdo.setattribute.php

**Pull request merged**